### PR TITLE
Rework exclusion handling to publish metrics

### PR DESCRIPTION
**Pull request by dlegaultbbry**
- Added excluded flags to some coverage metrics
- Changed code to set the flag instead of wiping things for branches, calls and conditions
- Adjusted reporting code to ignore now present but excluded items

Ran all docker tests, a few reference files were updated from the changes.

#988

### CODE REVIEW

File: gcovr/coverage.py [hunk 1]
```diff
@@ -201,18 +201,21 @@ class CallCoverage:
...
     """
 
-    __slots__ = "callno", "covered"
+    __slots__ = "callno", "covered", "excluded"
 
     def __init__(
         self,
         callno: int,
         covered: bool,
+        excluded: Optional[bool] = None,
```
Comment (Spacetown) [inline gcovr/coverage.py:214]: Use always a default value `False` (same as for `LineCoverage`).

File: gcovr/coverage.py [hunk 2]
```diff
@@ -230,17 +233,20 @@ class ConditionCoverage:
...
-    __slots__ = "count", "covered", "not_covered_true", "not_covered_false"
+    __slots__ = "count", "covered", "not_covered_true", "not_covered_false", "excluded"
 
     def __init__(
         self,
         count: int,
         covered: int,
         not_covered_true: List[int],
         not_covered_false: List[int],
+        excluded: Optional[bool] = None,
```
Comment (Spacetown) [inline gcovr/coverage.py:249]: Also default of `False`.

File: gcovr/coverage.py
Comment (Spacetown) [inline gcovr/coverage.py:1]: I would use the same functions as for the line coverage to the other coverage types:
```python
    @property
    def is_excluded(self) -> bool:
        """Return True if the line is excluded."""
        return self.excluded

    @property
    def is_reportable(self) -> bool:
        """Return True if the line is reportable."""
        return not self.excluded

    @property
    def is_covered(self) -> bool:
        """Return True if the line is covered."""
        return self.is_reportable and self.xxx

    @property
    def is_uncovered(self) -> bool:
        """Return True if the line is uncovered."""
        return self.is_reportable and not self.xxx
```

File: tests/excl-branch/reference/gcc-5/coverage.main.cpp.118fcbaaba162ba17933c7893247df3a.html [hunk 3]
```diff
@@ -116,9 +116,16 @@ <h1>GCC Code Coverage Report</h1>
...
+          <summary class="linebranchSummary">1/0</summary>
+          <div class="linebranchContents">
+            <div class="excludedBranch">&ndash; Branch 0 excluded.</div>
+            <div class="excludedBranch">&ndash; Branch 1 excluded.</div>
+          </div>
+        </details>
       </td>
-      <td class="linecount coveredLine">1</td>
-      <td class="src coveredLine"><span class="w">   </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="c1">// GCOVR_EXCL_BR_LINE</span></td>
+      <td class="linecount partialCoveredLine">1</td>
```
Comment (Spacetown) [inline tests/excl-branch/reference/gcc-5/coverage.main.cpp.118fcbaaba162ba17933c7893247df3a.html:127]: Why is this changed? I think this change shouldn't affect the reference files.

File: CHANGELOG.rst [hunk 4]
```diff
@@ -36,6 +36,7 @@ Internal changes:
 - Fix sporadic timestamp mismatch in development build package. (:issue:`1006`)
 - Replace ``black`` and ``flake8`` with ``ruff`` and move configuration of ``pytest`` to ``pyproject.toml``. (:issue:`1007`)
 - Add ``pylint`` for testing code. (:issue:`1014`)
+- Rework exclusion handling to allow publishing metrics (:issue:`1016`)
```
Comment (Spacetown) [inline CHANGELOG.rst:39]:
```suggestion
- Rework exclusion handling to keep information about excluded coverage. (:issue:`1016`)
```

**Review by Spacetown — CHANGES_REQUESTED** [review]
I think this PR should not affect any reference data because only the information is preserved and should be ignored for the report.

File: gcovr/exclusions/__init__.py [hunk 5]
```diff
@@ -141,9 +141,7 @@ def remove_internal_functions(filecov: FileCoverage):
                     linecov.function_name is not None
                     and linecov.function_name == function.name
                 ):
-                    linecov.excluded = True
-                    linecov.branches = {}
-                    linecov.count = 0
+                    linecov.exclude()
```
[the remark this answers is not in the document]
Comment (Spacetown) [reply gcovr/exclusions/__init__.py:144]: Than I suggest to exclude only branches and do the other things with the next PR.

**Comment by dlegaultbbry** [comment]
I'd propose we discuss the behavior before approving and doing the rest.

If I exclude a line, does that mean all branches, calls and conditions under it should report 0 as if never executed? This is what was happening earlier with the first patch set. Right now they are wiped from existence so the metrics never report them because they are deleted before the coverage output is produced. 

There is a slight inconsistency with calls and conditions are they are not emptied out like branches when the line is excluded. This happens in `remove_internal_functions` and `apply_exclusion_ranges`. Maybe it would be better to fix this inconsistency before making other changes to reduce the complexity of reviewing test diff changes.

**Comment by Spacetown** [comment]
I think it's better to fix the inconsistency first by also clearing the data and then continue on this PR which then shouldn't affect the test data.

**Review by Spacetown — COMMENTED** [review]
LGFM except the change in the noxfile.

**Review by Spacetown — APPROVED** [review]

File: noxfile.py [hunk 6]
```diff
@@ -404,6 +404,7 @@ def tests(session: nox.Session) -> None:
     set_environment(session)
     session.log("Print tool versions")
     session.run("python", "--version")
+    session.run("gcovr", "--version")
```
[the remark this answers is not in the document]
Comment (dlegaultbbry) [reply noxfile.py:407]: Wanted to see the gcovr version used printed when running tests. I found it useful, but I can revert the change.

[the remark this answers is not in the document]
Comment (Spacetown) [reply noxfile.py:407]: GCOVR is installed editable in the virtual environment, no need to print the version.

[the remark this answers is not in the document]
Comment (Spacetown) [reply noxfile.py:407]:
```suggestion
```

**Pull request merged**